src/osd/PGLog.h

```diff
@@ -88,13 +88,19 @@ struct PGLog : DoutPrefixProvider {
     mempool::osd_pglog::list<pg_log_entry_t>::reverse_iterator
       rollback_info_trimmed_to_riter;
 
+    /*
+     * return true if we need to mark the pglog as dirty
+     */
     template <typename F>
-    void advance_can_rollback_to(eversion_t to, F &&f) {
-      if (to > can_rollback_to)
-	can_rollback_to = to;
-
-      if (to > rollback_info_trimmed_to)
-	rollback_info_trimmed_to = to;
+    bool advance_can_rollback_to(eversion_t to, F &&f) {
+      bool dirty_log = to > can_rollback_to || to > rollback_info_trimmed_to;
+      if (dirty_log) {
+	if (to > can_rollback_to)
+	  can_rollback_to = to;
+
+	if (to > rollback_info_trimmed_to)
+	  rollback_info_trimmed_to = to;
```

Comment:
Instead of propagating dirty_log through return values, wouldn't it be easier to simply set it here?  Also, I think you need to set it to false in write_log_and_missing and write_log_and_missing_wo_missing.

Reply:
seems that the `advance_can_rollback_to` is a member of IndexedLog, but the `dirty_log` is the member of PGLog, so i means we can not set the `dirty_log` in `advance_can_rollback_to`, am i missing something? 

non static write_log_and_missing() will call undirty(), so maybe no need to clear the dirty_log?
static write_log_an_missing_wo_missing will not call undirty current, do we need to clear it?

Reply:
Ah, I missed the bit in undirty.